lib/commands/generate-json.js

```diff
@@ -0,0 +1,76 @@
+var fs              = require('fs');
+var path            = require('path');
+var mkdirp          = require('mkdirp');
+var Command         = require('../models/command');
+var JsonGenerator   = require('../utilities/json-generator');
+
+module.exports = Command.extend({
+  name: 'generate-json',
+  description: 'Generates a cache describing all commands and blueprints available for the command line',
+  aliases: [],
+  skipHelp: true,
+  works: 'everywhere',
+
+  availableOptions: [
+    { name: 'silent', type: Boolean, default: true, aliases: ['s'] },
+  ],
+
+  anonymousOptions: [],
+
+  cacheDirectory: '/node_modules/ember-cli',
+  cacheFileName: 'commands-help.json',
+
+  run: function(commandOptions, rawArgs) {
+
+    commandOptions.json = true;
+    var json = this._getJson(commandOptions, rawArgs);
+    var cache = JSON.stringify(json, function(key, value) {
+      var path = require('path');
+      // build command has a recursive property
+      if (value === path) {
+        return 'path';
+      }
+      return value;
+    }, 2);
+
+    if (rawArgs.length === 0) {
+      this._cacheJson(cache);
+    }
+
+    return cache;
+  },
+
+  _getJson: function(commandOptions, rawArgs) {
+    var generator = new JsonGenerator({
+      ui: this.ui,
+      project: this.project,
+      commands: this.commands,
+      tasks: this.tasks
+    });
+
+    return generator.generate(commandOptions, rawArgs);
+  },
+
+  _cacheJson: function(json) {
+
+    var directoryPath = path.join(__dirname, '../..')
```

Comment:
i actually changed it to `__dirname` for this exact reason. Here is the different behaviour:
```
$ /foobar -> npm link ember-cli
$ /foobar -> ember generate-json

this.project.root: /Users/lazybensch/src/opensource/foobar
__dirname: /Users/lazybensch/src/opensource/ember-cli/lib/commands
```

`__dirname` will **always** point to the correct location (since its that file that is executed it has to be in the ember-cli directory that is in charge right now)